arrow-optics/src/main/kotlin/arrow/optics/Fold.kt

```diff
@@ -38,44 +24,44 @@ interface Fold<S, A> : FoldOf<S, A> {
     /**
      * [Fold] that takes either [S] or [S] and strips the choice of [S].
      */
-    fun <S> codiagonal() = object : Fold<Either<S, S>, S> {
-      override fun <R> foldMap(M: Monoid<R>, s: Either<S, S>, f: (S) -> R): R = s.fold(f, f)
+    fun <S, R> codiagonal() = object : Fold<Either<S, S>, S> {
+      override fun <R> foldMap(s: Either<S, S>, empty: R, combine: (R, R) -> R, map: (S) -> R): R =
+        s.fold(map, map)
     }
 
     /**
      * Creates a [Fold] based on a predicate of the source [S]
      */
     fun <S> select(p: (S) -> Boolean): Fold<S, S> = object : Fold<S, S> {
-      override fun <R> foldMap(M: Monoid<R>, s: S, f: (S) -> R): R = if (p(s)) f(s) else M.empty()
+      override fun <R> foldMap(s: S, empty: R, combine: (R, R) -> R, map: (S) -> R): R =
+        if (p(s)) map(s) else empty
     }
 
     /**
      * [Fold] that points to nothing
      */
     fun <A, B> void() = POptional.void<A, B>().asFold()
 
-    /**
-     * Create a [Fold] from a [arrow.Foldable]
-     */
-    fun <F, S> fromFoldable(foldable: Foldable<F>) = object : Fold<Kind<F, S>, S> {
-      override fun <R> foldMap(M: Monoid<R>, s: Kind<F, S>, f: (S) -> R): R = foldable.run { s.foldMap(M, f) }
+    operator fun <S: Iterable<A>, A> invoke(): Fold<S, A> = object : Fold<S, A> {
+      override fun <R> foldMap(s: S, empty: R, combine: (R, R) -> R, map: (A) -> R): R =
+        s.map(map).fold(empty, combine)
     }
   }
 
   /**
    * Calculate the number of targets
    */
-  fun size(s: S) = foldMap(Int.monoid(), s = s, f = { _ -> 1 })
+  fun size(s: S) = foldMap(s, 0, Int::plus, { 1 })
 
   /**
    * Check if all targets satisfy the predicate
    */
-  fun forall(s: S, p: (A) -> Boolean): Boolean = foldMap(AndMonoid, s, p)
+  fun forall(s: S, p: (A) -> Boolean): Boolean = foldMap(s, true, Boolean::and, p)
 
   /**
    * Check if there is no target
    */
-  fun isEmpty(s: S): Boolean = foldMap(AndMonoid, s) { _ -> false }
+  fun isEmpty(s: S): Boolean = size(s) == 0
```

Comment:
How lazy are our folds? Do we have any lazy folds? If so we should re-encode this to short-circuit rather than check size which would walk the entire fold.